dynamiq/memory/backends/dynamo_db.py

```diff
@@ -0,0 +1,421 @@
+import json
+import time
+import uuid
+from decimal import Decimal, InvalidOperation
+from typing import Any
+
+from boto3.dynamodb.conditions import Attr
+from botocore.exceptions import ClientError
+from pydantic import ConfigDict, Field, PrivateAttr
+
+from dynamiq.connections import AWS
+from dynamiq.memory.backends.base import MemoryBackend
+from dynamiq.prompts import Message, MessageRole
+from dynamiq.utils.logger import logger
+
+
+class DynamoDBMemoryError(Exception):
+    """Base exception class for DynamoDB Memory Backend errors."""
+
+    pass
+
+
+def _convert_floats_to_decimals(obj):
+    if isinstance(obj, float):
+        if obj != obj or obj == float("inf") or obj == float("-inf"):
+            raise ValueError(f"Cannot convert non-finite float {obj} to Decimal for DynamoDB")
+        try:
+            return Decimal(str(obj))
+        except InvalidOperation:
+            raise ValueError(f"Could not convert float {obj} to Decimal")
+    elif isinstance(obj, dict):
+        return {k: _convert_floats_to_decimals(v) for k, v in obj.items()}
+    elif isinstance(obj, list):
+        return [_convert_floats_to_decimals(elem) for elem in obj]
+    else:
+        return obj
+
+
+class DynamoDB(MemoryBackend):
+    """
+    AWS DynamoDB implementation of memory storage using ONLY table scans.
+
+    Relies exclusively on DynamoDB Scan operations. Scans read the
+    entire table and can be slow and costly for large datasets.
+
+    Assumed Table Schema:
+    - PK: `message_id` (String)
+    - SK: `timestamp` (Number - Unix float stored as Decimal)
+    - Attributes: `role` (String), `content` (String), `metadata` (Map)
+    """
+
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+
+    name: str = "DynamoDB"
+    connection: AWS = Field(default_factory=AWS)
+    index_name: str = Field("conversations", description="Name of the DynamoDB table.")
+    create_table_if_not_exists: bool = Field(default=False)
+    billing_mode: str = Field(default="PAY_PER_REQUEST", pattern="^(PAY_PER_REQUEST|PROVISIONED)$")
+    read_capacity_units: int = Field(default=1, gt=0)
+    write_capacity_units: int = Field(default=1, gt=0)
+
+    partition_key_name: str = Field(default="message_id")
+    sort_key_name: str = Field(default="timestamp")
+    role_attribute_name: str = Field(default="role")
+    content_attribute_name: str = Field(default="content")
+    metadata_attribute_name: str = Field(default="metadata")
+
+    _dynamodb_resource: Any = PrivateAttr(default=None)
+    _dynamodb_table: Any = PrivateAttr(default=None)
+    _dynamodb_client: Any = PrivateAttr(default=None)
+
+    @property
+    def to_dict_exclude_params(self) -> dict[str, bool]:
+        return super().to_dict_exclude_params | {
```

Comment:
private attributes are not automatically included during serialization. Please recheck and remove them if this is True in such case.